Khatri-Rao product

jli05 opened the pull request:
* Add Khatri-Rao product API
* Fix `unittest.mk` for macOS

mli commented:
@JeanKossaifi can you please have a look at this PR?

JeanKossaifi commented:
A Khatri-Rao product is a great addition! My main problem is the misleading name of the function: what is the rational behind using the transposed matrices?

If it returns the transpose of the Khatri-Rao product then it is **not** a khatri-rao product - which is a **column-wise** Kronecker product. This is potentially very confusing the end user.

It should either return the actual Khatri-Rao product or be renamed (e.g. row_wise_kronecker).
As a side note an elegant way to do the tests would be to use the Kronecker product and compare with the appropriate bit.

JeanKossaifi commented on src/operator/contrib/krprod.h:
```diff
@@ -0,0 +1,138 @@
+ * \brief Computes Khatri-Rao product
+ *
+ * Given the transpose of input matrices, this function computes the
+ * transpose of the Khatri-Rao product of the input matrices.
```
Why the transpose?

jli05 replied:
It was mainly for the CP Decomposition, as the MXNet matrices are row-major, I found it easiest to store in and perform operations on effectively transposed factor matrices. But you're right the naming is confusing, I'll change it.

JeanKossaifi commented on src/operator/contrib/krprod.h:
```diff
@@ -0,0 +1,138 @@
+inline void krprod
+  (Tensor<cpu, 2, DType> out,
+  const std::vector<Tensor<cpu, 2, DType> > &ts_arr) {
+  // If no input matrix, return all-one vector
```
Why an all-one vector rather than, say, null?

jli05 replied:
This is so that the computation could carry on if the user calls `row_wise_kronecker()` (or `krprod()`) in a complex expression, e.g.

```c++
row_wise_kronecker(empty_ts_arr).dot(xxx)
```

JeanKossaifi replied:
Shouldn't we raise an error?

jli05 replied:
For example, NumPy has similar setting,

```python
In [2]: np.prod([2, 3, 5])
Out[2]: 30

In [3]: np.prod([])
Out[3]: 1.0
```

I thought returning all-one vector could be more tolerant and wouldn't crash the whole program.

JeanKossaifi replied:
I see the point, we might not want to raise an error (a warning maybe?) but I still thing it should at least be null or empty. Otherwise it doesn't make as much sense and we makes potential bugs in the code hard to find.
What do you think?

jli05 replied:
I also asked myself about cases that may make no sense. Could you give some concrete example of them just to clarify?

JeanKossaifi replied:
I can see cases where (especially if we make a Python binding to the function) users can accidentally pass an empty list (e.g. getting weight matrices corresponding to a neural net from a dict and not checking whether it is empty). Without explicit error this could lead to sneaky errors.
Conversely, I can't think of a case where we would **want** the khatri-rao product of nothing to be a vector of ones.

A happy medium could be to simply issue a warning and return a null result? The user could then deal with the null if needed?

jli05 replied:
I will enforce the input `ts_arr` to be non-empty, as I just wanted to handle the minimal set of cases for which we're sure of implemented logic.

JeanKossaifi commented on src/operator/contrib/krprod.h:
```diff
@@ -0,0 +1,138 @@
+    return;
+  }
+
+  // Check all input and output matrices have the same height
```
For khatri-rao they should have the same number of columns..
Also width and height is more for images, for matrices in general we might want to stick with row and columns.

jli05 commented:
@JeanKossaifi Could you take a look again?

JeanKossaifi commented on src/operator/contrib/krprod.h:
```diff
@@ -0,0 +1,202 @@
+  // Check all input and output matrices have the same number of rows
+  // and the output matrix has the right number of columns
+  int nrows = static_cast<int>(out.size(0));
+  int kronecker_length = 1;
```
Why does it say kronecker here?

jli05 replied:
These lines are within the function `row_wise_kronecker()`, so I named it `kronecker_length`.

JeanKossaifi replied:
I think `n_columns` is more appropriate and clearer but that's a detail :)

JeanKossaifi commented on src/operator/contrib/krprod.h:
```diff
@@ -0,0 +1,202 @@
+    expr::BLASEngine<cpu, DType>::SetStream
+      (result->stream_);
+
+    // Compute the current Khatri-Rao product, row by row
```
row-wise Kronecker rather than Khatri-Rao?

JeanKossaifi commented on src/operator/contrib/krprod.h:
```diff
@@ -0,0 +1,202 @@
+ * \param ts_arr vector of input matrices
+ */
+template <typename DType>
+inline void krprod
```
It seems suboptimal to transpose first and then to the product? 
Can we not do the operation directly on row-wise ordered arrays?

jli05 replied:
The Khatri-Rao product is by column. I don't quite see how to directly do it on the input matrices row-wise? ... Given that internally it'll necessarily wrap calls to LAPACK `<d>ger()` for outer product calculation (for efficiency), whose output is a "matrix", implying that at least the elements in its column will be contiguously stored.

JeanKossaifi replied:
The problem is that now we're copying everything first (when transposing). That might be slower than operating directly on the original matrices. Ideally we'd want to have a flag specifying whether the matrices need transposing or not (`gemm` style)

jli05 replied:
If we don't do transpose of matrices, we'll have to give up using LAPACK but write manually multi-fold loops to compute the Khatri-Rao product -- we may lose efficiency on this side.

JeanKossaifi replied:
Can we call it khatri-rao? krprod might be less clear for someone who doesn't know what to look for.

JeanKossaifi commented on include/mxnet/c_lapack_api.h:
```diff
@@ -54,6 +57,40 @@ extern "C" {
+ * \param lda leading dimension of a
+ */
+template <typename xpu, typename DType>
+inline void flip(int m, int n, DType *b, int ldb, DType *a, int lda);
```
mxnet already has transpose I believe: any reason to not use it?
https://github.com/dmlc/mxnet/blob/master/src/operator/tensor/matrix_op-inl.h

jli05 replied:
The `transpose()` defined in `mshadow/extension/transpose.h` operates on templated `Exp` and returns a templated `Exp` so that the compiler can finally instantiate out the entire Expression. What we're doing here is to *actually* transpose the layout of the data storage in memory so that later on LAPACK calls can read/interpret the data correctly.

For LAPACK libs that support only column-major matrices, we'd always need such a helper to transpose the data storage, perform the calculation, and transpose back the result. It seems the easiest way to handle row-major matrices and also the most efficient empirically.

JeanKossaifi replied:
Is there not a flag to specify whether the matrix needs to be transposed or to specify the ordering of the elements (like in gemm)?

jli05 replied:
The `flip()` does "flipping" of elements in the memory. A row-major matrix will be "transposed" if we consider the result matrix still following the row-major layout; or alternatively, the result matrix will be "the same" as the input one if we consider the result matrix following column-major layout. Similar goes for a column-major input matrix.

Therefore there's no flag for the layout in this function's signature. All it does is mechanical "flipping" of elements. It's left to the developer how to interpret and use the result.

JeanKossaifi replied:
I was asking for the LAPACK function, was hoping there would be a flag to specify whether any of the inputs needs to be transposed rather than having to do it explicitly ourselves beforehand..

jli05 replied:
This function is intended to be a helper function that just does "flipping" or transpose of elements in memory, for the use of other MXNet LAPACK interfaces. It sits beneath those interfaces.

By the design in `include/mxnet/c_lapack_api.h`, all MXNet LAPACK interfaces need to be able to handle both row-major and column-major matrices. When we're building against LAPACK libs that only support column-major matrices natively, we'll have to provide support for row-major ones. Sometimes we could do it cleverly as in `dgesdd` (SVD), but many more times we do have to transpose some input matrix under the hood before feeding to the native LAPACK routine, e.g. `dposv` (linear solver).

People may argue that for many functions, we could do things in a theoretically equivalent way without having to do hard transpose of input matrix. I did some test -- we may lose on numerical accuracy in the general case. It seems to me that when we do need it, transposing prior and after the native call seems an easy-to-implement, numerically conservative option.

JeanKossaifi commented:
Thanks for the changes @jli05 !

As a side note, mxnet seems to have some blas/lapack interfaces [in here](https://github.com/dmlc/mxnet/blob/master/src/operator/tensor/la_op_inline.h), would it make sense to add the actual operator there too? Any thoughts @mli ?

jli05 commented:
I think it'd be a good idea to add Operator in the file you pointed out. Could I do it later on? I was just aiming to provide C++ primitives in this PR.

jli05 commented:
@JeanKossaifi I did fixes you pointed out and also

* added linear solver for positive, semi-definite l.h.s matrix A (`MXNET_LAPACK_posv`), which handles both row-major & column-major matrices for the r.h.s matrix B
* added the Moore-Penrose pseudo inverse of the Khatri-Rao product

Could you take a look?

JeanKossaifi commented on src/operator/contrib/krprod.h:
```diff
@@ -0,0 +1,290 @@
+}
+
+/*!
+ * \brief Moore-Penrose pseudoinverse of the Khatri-Rao product
```
This function could probably use more comments especially on how the inverse is done for people who might want to modify in the future :)

JeanKossaifi replied:
Great with the new comments!

Does it return the pseudo inverse of the khatri-rao product or the **Transpose** of the pseudo inverse of the kr product?

jli05 replied:
It always returns the pseudo inverse of the Khatri-Rao product, never the transpose of the pseudo inverse.

JeanKossaifi replied:
Just a side thought then: if we are working on the khatri-rao of the transpose (which we nedd e.g. for CP decomposition as you mentioned), we are in a way working with row-wise kronecker product and we can now avoid all the transposition: maybe something to keep in mind for when we do that?

jli05 replied:
Yes I'm going to submit the refactored `CPDecomp` in another PR. We'll see we generally store and operator on transposed factor matrices, which free ourselves from doing transpose.

JeanKossaifi commented on src/operator/contrib/krprod.h:
```diff
@@ -0,0 +1,290 @@
+
+/*!
+ * \brief Moore-Penrose pseudoinverse of Khatri-Rao product of two matrices
+ *
```
Any reason to have this function in addition to the general one?

jli05 replied:
For CP Decomposition, as

unfolded_tensor = A_i \Phi (A_n \khatri-rao A_{n-1} ... \khatri-rao A_1)^T

with `inv_krprod()` we could compute the last term on the r.h.s by one call and make the implementation of ALS much easier and straightforward.

JeanKossaifi replied:
I meant why do we need the function for two matrices when we have the function for an arbitrary number of matrices?

jli05 replied:
This is just a convenience function that makes the notation a bit shorter if we're computing on two input matrices. It's not absolutely mandatory if you want.

JeanKossaifi replied:
I like the explicit name, I think it might be confusing to have two similar functions but I do not have a strong opinion :)

jli05 commented:
@JeanKossaifi Could you check again?

JeanKossaifi commented:
Thanks for all the great work @jli05, looks good to me!

I generally favor explicit names so would go for `khatri_rao` and `inv_khatri_rao` instead of `krprod` and `inv_kprod` but that's more or less it :)

jli05 commented:
You're right @JeanKossaifi . I renamed those functions and removed all convenience functions.

jli05 commented:
@mli May we merge?

jli05 commented:
@mli What do you think we can further do on this PR? May we consider merging?

domdivakaruni commented:
this is cool! Thanks @jli05. @mli can we merge?

piiswrong commented on include/mxnet/c_lapack_api.h:
```diff
@@ -70,6 +113,38 @@ inline char loup(char uplo, bool invert) { return invert ? (uplo == 'U' ? 'L' :
   MXNET_LAPACK_CWRAPPER1(spotri, float)
   MXNET_LAPACK_CWRAPPER1(dpotri, double)
 
+  inline int MXNET_LAPACK_sposv(int matrix_layout, char uplo, int n, int nrhs,
```
mxnet_lapack_sposv

jli05 commented:
@piiswrong could you check again?

jli05 commented:
@piiswrong what do you think about this PR? May we merge?
